src/contracts/mod.rs

```diff
@@ -0,0 +1,227 @@
+            }
+        } else {
+            info!(group_id = ?options.group_id, ?tree_depth, "Semaphore group found.");
+        }
+
+        // TODO: Some way to check the initial leaf
+
+        Ok(Self {
+            ethereum,
+            semaphore,
+            group_id: options.group_id,
+            tree_depth: usize::from(tree_depth),
+            initial_leaf: options.initial_leaf,
+        })
+    }
+
+    #[must_use]
+    pub const fn group_id(&self) -> U256 {
+        self.group_id
+    }
+
+    #[must_use]
+    pub const fn tree_depth(&self) -> usize {
+        self.tree_depth
+    }
+
+    #[must_use]
+    pub const fn initial_leaf(&self) -> Field {
+        self.initial_leaf
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    pub async fn fetch_events(
+        &self,
+        starting_block: u64,
+        last_leaf: usize,
+    ) -> EyreResult<Vec<(usize, Field, Field)>> {
+        info!(starting_block, "Reading MemberAdded events from chains");
+        // TODO: Register to the event stream and track it going forward.
+
+        // Fetch MemberAdded log events
+        let filter = self
+            .semaphore
+            .member_added_filter()
+            .from_block(starting_block);
+        let events = self
+            .ethereum
+            .fetch_events::<MemberAddedEvent>(&filter.filter)
+            .await?;
+
+        info!(count = events.len(), "Read events");
+        let mut index = last_leaf;
+        let insertions = events
+            .iter()
+            .filter(|event| event.group_id == self.group_id)
+            .map(|event| {
+                let mut id_bytes = [0u8; 32];
+                event.identity_commitment.to_big_endian(&mut id_bytes);
+
+                let mut root_bytes = [0u8; 32];
+                event.root.to_big_endian(&mut root_bytes);
+
+                // TODO: Check for < Modulus.
+                let root = Field::from_be_bytes_mod_order(&root_bytes);
+                let leaf = Field::from_be_bytes_mod_order(&id_bytes);
+                let res = (index, leaf, root);
+                index += 1;
+                res
+            })
+            .collect::<Vec<_>>();
+        Ok(insertions)
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    pub async fn is_manager(&self) -> EyreResult<bool> {
+        info!(address = ?self.ethereum.address(), "My address");
+        let manager = self.semaphore.manager().call().await?;
+        info!(?manager, "Fetched manager address");
+        Ok(manager == self.ethereum.address())
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    pub async fn insert_identity(&self, commitment: &Field) -> EyreResult<()> {
+        info!(%commitment, "Inserting identity in contract");
+
+        // Send create tx
+        let commitment = U256::from(commitment.to_be_bytes());
+        self.ethereum
+            .send_transaction(self.semaphore.add_member(self.group_id, commitment).tx)
```

Comment:
We probably need to specify an explicit gas limit, I think the estimate fails, which leads to following error:
```
ERROR signup_sequencer::ethereum: Transaction failed, e: Send(MiddlewareError(MiddlewareError(MiddlewareError(MiddlewareError(JsonRpcClientError(JsonRpcError(JsonRpcError { code: -32000, message: "gas required exceeds allowance (5)", data: None })))))))
```

Reply:
Let's investigate this more than. If it's more than the +10% + 10k (which it uses now), then I'd like to know the actual numbers and know where the delta is coming from.